rust/frontend/src/function.rs

```diff
@@ -0,0 +1,476 @@
+//! This module provides idiomatic Rust API for creating and working with TVM functions.
+//!
+//! For calling an already registered TVM function use [`function::Builder`] and to register
+//! a TVM packed function from Rust either use [`function::register`] or the
+//! macro [`register_global_func`].
+//!
+//! See the `tests` and `examples` repository for usage examples.
+
+use std::ffi::{CStr, CString};
+use std::mem;
+use std::os::raw::{c_char, c_int, c_void};
+use std::ptr;
+use std::slice;
+use std::str;
+use std::sync::Mutex;
+
+use ts;
+
+use ty::TypeCode;
+use value::TVMValue;
+use value::ValueKind;
+use Error;
+use Module;
+use Result;
+use TVMArgValue;
+use TVMRetValue;
+
+lazy_static! {
+    static ref GLOBAL_FUNCTION_NAMES: Mutex<Vec<&'static str>> = list_global_func_names();
+}
+
+fn list_global_func_names() -> Mutex<Vec<&'static str>> {
+    let mut out_size = 0 as c_int;
+    let mut name = ptr::null() as *const c_char;
+    let mut out_array = &mut name as *mut _;
+    check_call!(ts::TVMFuncListGlobalNames(
+        &mut out_size as *mut _,
+        &mut out_array as *mut _
+    ));
+    let list = unsafe { slice::from_raw_parts(out_array, out_size as usize) };
+    let list = list
+        .iter()
+        .map(|&p| unsafe { CStr::from_ptr(p) })
+        .map(|cs| cs.to_bytes())
+        .map(|bs| str::from_utf8(bs).unwrap())
+        .collect();
+    Mutex::new(list)
+}
+
+/// Returns a registered TVM function by name.
+pub fn get_global_func(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+    let name = CString::new(name).unwrap();
+    let mut handle = ptr::null_mut() as ts::TVMFunctionHandle;
+    check_call!(ts::TVMFuncGetGlobal(
+        name.as_ptr() as *const c_char,
+        &mut handle as *mut _
+    ));
+    if !(handle.is_null()) {
+        mem::forget(name);
+        return Some(Function::new(handle, is_global, false));
+    } else {
+        if allow_missing {
+            return None;
+        } else {
+            panic!("Cannot find global function {:?}", name);
+        }
+    }
+}
+
+/// Wrapper around TVM function handle.
+#[derive(Debug, Hash)]
+pub struct Function {
+    handle: ts::TVMFunctionHandle,
+    is_global: bool,
+    is_released: bool,
```

Comment:
Here struct fields are private and cargo doc ignores them. But I'll keep them as comment about them and add doc in the Function doc.

Reply:
yes, please. `is_global` is clear but `is_released` can be confusing if someone comes from a background where `free` isn't synonymous with `release`. I'm almost tempted to guess you have some objective-C experience!

Reply:
No, fortunately I don't have any objective-C experience :) I used some of existing java and python frontend namings. I'll change it to `is_free` later.

Reply:
no, no `is_released` is quite fine! a short comment saying

```
// whether the tensor has been freed by the TVM runtime
```

would suffice

Reply:
Guess the more idiomatic would be `is_dropped`.

Reply:
hmm, but the term `dropped` means something very specific in Rust. I would expect that to be related to the `Drop` trait but that's misleading since it's whether the TVM dylib has "dropped" the object. I'd do

```
is_released
is_freed
```

or, alternatively, set the function handle to `ptr::null` and use that as a sentinel value for being dropped